Added remaining rulesets, cleaned up stale configuration and analyzers and added documentation around banned methods

jryckman opened the pull request:
# Justification
This change does several things:
- Adds remaining rulesets that we want to be a part of our package
- Removes LiteralExemptions as it's a stale rule and needs to be removed from our analyzer
- Updates banned methods to include documentation on how to use it, but to also not include a preset one in our package. This is a better approach because banned methods should live in the consuming project's source control as opposed to our package, to mirror what we did with CodeAnalysisDictionary.xml
- Update NI.Analysis.targets to include NI.RoslynTransition.targets instead of having separate files, since it's not clear we're ever going to be able to remove it. I also added more TODO and documentation in the file to explain what is going on and what needs to change before we can package everything up.

# Implementation
I moved files around to be in slightly different locations (configuration directory), but that doesn't actually change anything until we package things up. For the particular path changes, I've updated spots in the NI.Analysis.targets file to point to the new, relative path.
There are some lines in the .targets file that were deleted and added, which I'll try to add comments around to identify the key changes vs. straight copy behavior.

# Testing
No testing this round since it's all configuration and no new analyzers. Relying on existing CI.

jryckman committed:
Create first pass of analyzers and tests that build and all unit tests pass

jryckman committed:
Collapse Analyzers.Naming into Analyzers to consolidate. Moved tests and analyzer code, consolidated namespaces and deleted existing projects.

jryckman committed:
Rename assemblies and folders to match. Update namespaces to get all code compiling and tests passing

jryckman committed:
Renamed Everything.sln to NationalInstruments.Analyzers.sln.
Renamed test folder to tests to match dotnet github project

jryckman committed:
Delete packages.config as it wasn't necessary and cleaned up nuget.config

jryckman committed:
Add content from the analyzer wrapper package user branch. Configured these projects to use NI.ruleset and fixed warnings.
NOTE: There's a chicken-and-egg problem because we want to also consume the analyzers that we build here as a part of the ruleset, but we need to get an initial pass working first.
Once we build a package of our analzyers, we will start consuming that packge on newer versions of itself to ensure we're also eating our own dog food.

jryckman committed:
Merge remote-tracking branch 'upstream/main' into main

jryckman committed:
Removing files not required to make release builds pass

jryckman committed:
Created documentation on using banned methods

Created a .md file to explain how to use the Banned Methods analyzer rule

jryckman committed:
Update and remove assets that shouldn't be included

LiteralExemptions is something that is no longer being used, so we should remove it from the source/package. I moved the rest of the ruleset files that we want to have into the rulesets directory.

I moved the .targets file and additional files into a configuration directory that are used to configure the analyzers to run properly. This can then get packaged into a content directory or something else.

jryckman committed:
Merge remote-tracking branch 'upstream/main' into main

jryckman commented on configuration/NI.Analysis.targets:
```diff
@@ -0,0 +1,107 @@
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+
+  <PropertyGroup>
+    <!-- TODO: This location of 'content' needs to get updated once we build our package if we decide to put the files elsewhere, outside of a 'content' directory -->
```
I added this TODO to note that we may need to update this once we create our new package. If it stays the same, we should just remove the TODO.

pakdev replied:
Do we expect clients to define `PkgNI_DotNet_Analyzers`? If so, I'd just get rid of "content"

jryckman commented on configuration/NI.Analysis.targets:
```diff
@@ -0,0 +1,107 @@
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <NI1006_BannedMethods>$(SolutionDir)\BannedMethods.xml</NI1006_BannedMethods>
```
I removed Literal Exemptions from this file since it won't be used anymore. I changed BannedMethods to be included from $SolutionDir instead of anything relative to this package.

jryckman commented on configuration/NI.Analysis.targets:
```diff
@@ -0,0 +1,107 @@
+  <!-- Disable portions of Roslyn analyzers that would otherwise generate errors in our build. There should be a goal
+       around removing as many of these as possible, but we want to report warnings and elevate warnings as errors in
+       our release builds rather than erroring all the time on everything. -->
+  <ItemGroup>
```
This is the content that was in NI.RoslynTransition.targets file that I'm just putting directly in here. I don't see a reason to maintain a separate file, but we could if people feel strongly about it.

DelpireNI replied:
I think I could go either way here. It doesn't quite seem appropriate in this file, but its so little, it doesn't seem worth worrying about adding here. I think if we expect this to grow significantly, it would be nice to have as a separate file.

jryckman replied:
Yeah I agree. This is the largest it's every going to be, and it may shrink as time goes on, though I'm not sure by how much

jryckman committed:
Added back BannedMethods additional files

Added additional file for banned methods to ensure it's picked up by the analyzer

pakdev commented on configuration/NI.Analysis.targets:
```diff
@@ -0,0 +1,110 @@
+    <NuGetPackagesDirectory>$(NuGetPackageRoot)</NuGetPackagesDirectory>
+  </PropertyGroup>
+
+	<PropertyGroup>
```
Any reason why this is extra indented? Seems like the children aren't 2 spaces either

jryckman replied:
no, I'm not sure why. I'll clean that up and push an update

alejandro5042 commented on configuration/NI.Analysis.targets:
```diff
@@ -0,0 +1,110 @@
+    <NuGetPackagesDirectory>$(NuGetPackageRoot)</NuGetPackagesDirectory>
+  </PropertyGroup>
+
+	<PropertyGroup>
```
Mix of tabs and spaces in this file :)

Replace with spaces, 2 or 4, preferably 4.

jryckman replied:
Replaced with 2 spaces which matches our ASW conventions for XML config files.

alejandro5042 commented on docs/Configuring Banned Methods.md:
```diff
@@ -0,0 +1,46 @@
+# Process
+
+The analyzer works by looking through "BannedMethods" [additional files](https://github.com/dotnet/roslyn/blob/master/docs/analyzers/Using%20Additional%20Files.md). A project can contain any number of files as long as:
+
```
Double blank line, remove?

jryckman replied:
If I didn't put in two blank lines, then I couldn't get the extra newline I wanted in the preview of the MD.

alejandro5042 commented on rulesets/NI.TestUtilities.ruleset:
```diff
@@ -0,0 +1,156 @@
+    <Rule Id="CA1709" Action="None" />
+    <Rule Id="CA1710" Action="None" />
+    <Rule Id="CA1711" Action="None" />
+    <Rule Id="CA1712" Action="None" /> <!-- RoslynTransition: Should be "Warning". Tracked by Dev Efficiency: https://trello.com/b/gtoe4Of2/iteration-board -->
```
Should we track this a different way before posting this on GitHub? Who should be tracking this?

jryckman replied:
We should actually probably just remove all of that.  At this point I think it's stale, and most of these are

alejandro5042 commented on rulesets/NI.Tests.ruleset:
```diff
@@ -0,0 +1,156 @@
+    <Rule Id="CA1709" Action="None" />
+    <Rule Id="CA1710" Action="None" />
+    <Rule Id="CA1711" Action="None" />
+    <Rule Id="CA1712" Action="None" /> <!-- RoslynTransition: Should be "Warning". Tracked by Dev Efficiency: https://trello.com/b/gtoe4Of2/iteration-board -->
```
Should we track this a different way before posting this on GitHub? Who should be tracking this?

jryckman replied:
I am deleting all of these. Almost all of them are stale and we should just track everything as we find it from here on out in github or at least in the coding practices chapter backlog.

alejandro5042 commented on configuration/NI.Analysis.targets:
```diff
@@ -0,0 +1,110 @@
+  </PropertyGroup>
+
+	<PropertyGroup>
+		<!-- this would more correctly be: assemblyname contains .Test. We have an issue with test utilities projects, and test.testutilities projects. -->
```
We should document the behavior in these properties somewhere in the README. That way people who use these rules know when some rules apply and others don't.

jryckman replied:
Yes, I agree that we need to add more documentation to the README. I'll make an action item to do that in a followup submission. I already have a work item for someone to help with the README in general

jryckman replied:
I have a task in my backlog to add documentation on the different rulesets and which to use when and what behaviors are different.

alejandro5042 commented on configuration/NI.Analysis.targets:
```diff
@@ -0,0 +1,110 @@
+
+	<PropertyGroup>
+		<!-- this would more correctly be: assemblyname contains .Test. We have an issue with test utilities projects, and test.testutilities projects. -->
+		<IsNITestProject Condition="'$(IsNITestProject)' != 'True'">$(MSBuildProjectName.StartsWith("Tests."))</IsNITestProject>
```
Minor: Consider renaming `IsNITestProject` to `NI_IsTestProject`... cause non-NI people could theoretically use this style guide and NI doesn't claim those projects :)

jryckman replied:
Renamed

alejandro5042 commented on configuration/NI.Analysis.targets:
```diff
@@ -0,0 +1,110 @@
+
+	<PropertyGroup>
+		<!-- this would more correctly be: assemblyname contains .Test. We have an issue with test utilities projects, and test.testutilities projects. -->
+		<IsNITestProject Condition="'$(IsNITestProject)' != 'True'">$(MSBuildProjectName.StartsWith("Tests."))</IsNITestProject>
```
Ideally, properties that these `.targets`/`.props` files introduce should be prefixed with something like `NI_` or `NI_CSharpStyleGuide_` so that they don't collide with the user's properties.

jryckman replied:
Good idea. I went with NI_, and updated NI_IsTestProject and NI_IsTestUtilitiesProject

jryckman committed:
Fix tabs vs. whitespacing issue.

pakdev commented on configuration/NI.Analysis.targets:
```diff
@@ -0,0 +1,110 @@
+    <NuGetPackagesDirectory>$(NuGetPackageRoot)</NuGetPackagesDirectory>
+  </PropertyGroup>
+
+	<PropertyGroup>
```
(Optional) Perhaps we should put a condition on this e.g. `<PropertyGroup Condition="'$(IsNITestProject)' == ''">`. That way, if a project is being classified incorrectly, it can fix itself by manually setting `<IsNITestProject>true|false</IsNITestProject>`. Then, I'd add another PropertyGroup for `IsNITestUtilitiesProject` and one for the ruleset properties (or move them into another).

jryckman replied:
I like this idea. I'm going to do this in a follow-up submission where we have a good diff on this one. Adding to backlog.

pakdev commented on configuration/NI.Analysis.targets:
```diff
@@ -0,0 +1,110 @@
+
+  <PropertyGroup>
+    <!-- Enable Roslyn analyzers if they haven't been explicitly disabled, we're using Roslyn for CA/SA, the code needs to be validated -->
+    <EnableRoslynAnalyzers>True</EnableRoslynAnalyzers>
```
This will always enable them. You need to add a condition like `<EnableRoslynAnalyzers Condition="'$(EnableRoslynAnalyzers)' == ''">true</EnableRoslynAnalyzers>`

jryckman replied:
After talking with Peter, I'm consolidating the three properties into a single one, NI_EnableAnalyzers and I updated everything accordingly

pakdev commented on configuration/NI.Analysis.targets:
```diff
@@ -0,0 +1,110 @@
+     <Link>$(CodeAnalysisRuleSet)</Link>
+    </AdditionalFiles>
+
+    <AdditionalFiles Include="$(NI1704_GlobalExceptions)" Condition="Exists('$(NI1704_GlobalExceptions)')" />
```
Perhaps it's too big, but just confirming you didn't want a `<Link />` here too

jryckman replied:
I assume we want a link there. I'll add it.

jryckman committed:
Update .editorconfig to have indent size for XML config files

Update .editorconfig to have indent size for XML config files

pakdev commented on configuration/NI.Analysis.targets:
```diff
@@ -0,0 +1,110 @@
+    <NoWarn>$(NoWarn),@(RoslynTransition_DisabledRule)</NoWarn>
+  </PropertyGroup>
+  <ItemGroup>
+    <HashFiles Include="$(MSBuildThisFileFullPath)" />
```
ASW uses this for cache invalidation. Rather than assume all projects do, I think we should remove this and let ASW add e.g. `.nuget\packages.config` to HashFiles

jryckman replied:
I've removed the hash item. We'll need to remember to update that in ASW though at some point

pakdev commented on docs/Configuring Banned Methods.md:
```diff
@@ -0,0 +1,46 @@
+
+# Process
+
+The analyzer works by looking through "BannedMethods" [additional files](https://github.com/dotnet/roslyn/blob/master/docs/analyzers/Using%20Additional%20Files.md). A project can contain any number of files as long as:
```
"This analyzer works by loading banned methods from [additional files](url). An additional file's contents will be parsed if the following conditions are true:"

jryckman replied:
Nice, thanks! Updated.

pakdev commented on docs/Configuring Banned Methods.md:
```diff
@@ -0,0 +1,46 @@
+
+
+1. The filename contains the text "BannedMethods"
+2. The contents of the file are XML with the appropriate schema
```
Get rid of 2 (or you can add a note that an malformed additional file will output a diagnostic)

jryckman replied:
I'll get rid of it. I talk about the schema below, and I think that's fine

pakdev commented on rulesets/NI.TestUtilities.ruleset:
```diff
@@ -0,0 +1,156 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<RuleSet Name="NI.Tests.ruleset" Description="Code analysis rule overrides for NI Test projects." ToolsVersion="14.0">
```
Name needs to be updated

jryckman replied:
Updated

jryckman committed:
Rename properties for NI prefixing

Renamed NI_IsTestProject and NI_IsTestUtilitiesProject

pakdev reviewed (approved):
Consider some of my comments and fix others

jryckman committed:
Prefix properties with NI if we defined them

Prefix properties with NI if we defined them

jryckman committed:
Update documentation on banned methods

Update documentation on banned methods

jryckman committed:
Update banned method documentation

Removed #2 from the conditions.

jryckman committed:
Address review comments, add link and rename ruleset

Added a link to the global exceptions and updated the <RuleSet> to be match the file.

jryckman committed:
Remove stale comments

Remove stale comments
